# Bonus task

In [1]:
import nltk
nltk.download('brown')
from nltk.corpus import brown
from collections import Counter
import re
import math

[nltk_data] Downloading package brown to
[nltk_data]     C:\Users\Emmanuella\AppData\Roaming\nltk_data...
[nltk_data]   Package brown is already up-to-date!


In [2]:
def normalise_words(tokens):
    return [t.lower() for t in tokens if re.fullmatch(r"[A-Za-z]+(?:'[A-Za-z]+)?", t)]

words = normalise_words(brown.words())

In [3]:
def get_counts(words):
    word_counts = Counter(words)
    bigram_count = Counter(zip(words[:-1], words[1:]))
    return word_counts, bigram_count

def compute_pmi(words, min_freq=10):
    word_count, bigram_count = get_counts(words)

    valid_freq_words = {w for w, c in word_count.items() if c >= min_freq}
    
    filtered_bigrams = {(w1, w2): c for (w1, w2), c in bigram_count.items() if w1 in valid_freq_words and w2 in valid_freq_words}

    N_words = len(words)
    N_bigrams = len(words)-1
    scores = []

    for (w1, w2), c12 in filtered_bigrams.items():
        p_w1 = word_count[w1] / N_words
        p_w2 = word_count[w2] / N_words
        p_w1w2 = c12 / N_bigrams

        pmi = math.log2(p_w1w2 / (p_w1*p_w2))
        scores.append(((w1, w2), pmi, c12))

    return sorted(scores, key=lambda x: x[1])


In [4]:
pmi_scores = compute_pmi(words, min_freq=10)

print("Lowest 20 PMI pairs:")
for pair, score, count in pmi_scores[:20]:
    print(pair, score, count)

print("\nHighest 20 PMI pairs:")
for pair, score, count in reversed(pmi_scores[-20:]):
    print(pair, score, count)

Lowest 20 PMI pairs:
('the', 'not') -8.345869136255875 1
('the', 'but') -8.272362600262126 1
('with', 'of') -8.064480161546472 1
('an', 'the') -8.044140655714173 1
('his', 'of') -8.005289550382798 1
('the', 'his') -7.947632674860827 2
('the', 'you') -7.857434865889249 1
('a', 'he') -7.8033576879376 1
('the', 'it') -7.687070754706554 3
('the', 'would') -7.581523106358029 1
('i', 'the') -7.509591386246365 2
('he', 'of') -7.453957827542922 2
('a', 'his') -7.3546894107774765 1
('his', 'a') -7.3546894107774765 1
('the', 'from') -7.268735665327135 2
('the', 'no') -7.238038868104219 1
('of', 'of') -7.215176340361353 9
('the', 'out') -7.209429247035767 1
('the', 'the') -7.140504573814757 35
('the', 'a') -7.091881825955217 12

Highest 20 PMI pairs:
('hong', 'kong') 16.459719770244078 11
('viet', 'nam') 15.919151388881374 13
('simms', 'purdew') 15.831688547631035 12
('pathet', 'lao') 15.831688547631035 10
('herald', 'tribune') 15.56663497416059 7
('lo', 'shu') 15.459719770244076 21
('el', 'paso'

In [5]:
#extended version for step 3
def compute_pmi(words, min_freq=10, positive=False):
    word_count, bigram_count = get_counts(words)

    valid_freq_words = {w for w, c in word_count.items() if c >= min_freq}
    
    filtered_bigrams = {(w1, w2): c for (w1, w2), c in bigram_count.items() if w1 in valid_freq_words and w2 in valid_freq_words}

    N_words = len(words)
    N_bigrams = len(words)-1
    scores = []

    for (w1, w2), c12 in filtered_bigrams.items():
        p_w1 = word_count[w1] / N_words
        p_w2 = word_count[w2] / N_words
        p_w1w2 = c12 / N_bigrams

        pmi = math.log2(p_w1w2 / (p_w1*p_w2))
        if positive:
            pmi = max(0, pmi)

        scores.append(((w1, w2), pmi, c12))

    return sorted(scores, key=lambda x: x[1])

In [7]:
with open("brown_100.txt", "r") as f:
    brown100_text = f.read()

brown100_text = brown100_text.replace("<s>", " ").replace("</s>", " ")
brown100_words = normalise_words(brown100_text.split())

brown_pmi = compute_pmi(words, min_freq=10, positive=False)
brown_ppmi = compute_pmi(words, min_freq=10, positive=True)

brown100_pmi = compute_pmi(brown100_words, min_freq=10, positive=False)
brown100_ppmi = compute_pmi(brown100_words, min_freq=10, positive=True)

# PMI RESULTS (Brown corpus) 
print("Lowest 20 PMI pairs (Brown):")
for pair, score, count in brown_pmi[:20]:
    print(pair, score, count)

print("\nHighest 20 PMI pairs (Brown):")
for pair, score, count in reversed(brown_pmi[-20:]):
    print(pair, score, count)

# PMI RESULTS (brown_100) 
print("\nLowest 20 PMI pairs (Brown 100):")
for pair, score, count in brown100_pmi[:20]:
    print(pair, score, count)

print("\nHighest 20 PMI pairs (Brown 100):")
for pair, score, count in reversed(brown100_pmi[-20:]):
    print(pair, score, count)

# PPMI RESULTS (Brown corpus)
print("\nTop 20 PPMI pairs (Brown):")
for pair, score, count in reversed(brown_ppmi[-20:]):
    print(pair, score, count)

print("\nLowest 20 PPMI pairs (Brown):")
for pair, score, count in brown_ppmi[:20]:
    print(pair, score, count)

# PPMI RESULTS (brown_100) 
print("\nTop 20 PPMI pairs (Brown 100):")
for pair, score, count in reversed(brown100_ppmi[-20:]):
    print(pair, score, count)

print("\nLowest 20 PPMI pairs (Brown 100):")
for pair, score, count in brown100_ppmi[:20]:
    print(pair, score, count)

Lowest 20 PMI pairs (Brown):
('the', 'not') -8.345869136255875 1
('the', 'but') -8.272362600262126 1
('with', 'of') -8.064480161546472 1
('an', 'the') -8.044140655714173 1
('his', 'of') -8.005289550382798 1
('the', 'his') -7.947632674860827 2
('the', 'you') -7.857434865889249 1
('a', 'he') -7.8033576879376 1
('the', 'it') -7.687070754706554 3
('the', 'would') -7.581523106358029 1
('i', 'the') -7.509591386246365 2
('he', 'of') -7.453957827542922 2
('a', 'his') -7.3546894107774765 1
('his', 'a') -7.3546894107774765 1
('the', 'from') -7.268735665327135 2
('the', 'no') -7.238038868104219 1
('of', 'of') -7.215176340361353 9
('the', 'out') -7.209429247035767 1
('the', 'the') -7.140504573814757 35
('the', 'a') -7.091881825955217 12

Highest 20 PMI pairs (Brown):
('hong', 'kong') 16.459719770244078 11
('viet', 'nam') 15.919151388881374 13
('simms', 'purdew') 15.831688547631035 12
('pathet', 'lao') 15.831688547631035 10
('herald', 'tribune') 15.56663497416059 7
('lo', 'shu') 15.459719770244076 

The word pairs with the lowest PMI are mostly function words that sound wrong or unusual next to each other, such as ('the', 'not') at -8.35 and ('the', 'the') at -7.14. This shows that when the previous word is 'the', some next words become much less likely, which goes against the idea that words are independent. The highest PMI pairs are mostly names and fixed phrases, such as ('hong', 'kong') at 16.46 and ('viet', 'nam') at 15.92, where the two words are very strongly linked and usually appear together. The top PPMI results are the same as the PMI results because those scores were already positive. At the bottom, however, all negative PPMI scores become 0, so it loses information about how strongly word pairs avoid each other. For example, ('which', 'the') with 245 occurrences and ('thanks', 'of') with only 1 occurrence both end up with the same PPMI score of 0.

For the Brown 100 corpus, the words that remain after the min_freq=10 filter are mostly topic-specific words from this news article about the Fulton County election. The highest PMI pair is ('fulton', 'county') at 5.80, which is much lower than the full corpus maximum of 16.46. This is because rare but very specific name pairs like ('hong', 'kong') do not appear in this smaller dataset. The lowest PMI pairs are again unusual function-word combinations, such as ('said', 'the') at -0.94, but the negative scores are much smaller than in the full corpus because the smaller amount of data gives weaker statistical evidence. PPMI behaves the same way as in the full corpus: positive pairs keep the same values as PMI, while all negative pairs are reduced to 0.